File: fmt/cppfmt/src/doubleintegrator.cpp
#include "doubleintegrator.h"
#include <math.h>
#include "matplotlibcpp.h"

using namespace Eigen;

double bisection_newton(pf_cost f, pf_cost df, double left, double right, double eps, double itr_max)
{
    double x_est = (left + right) / 2;
    for (unsigned int i = 0; i < itr_max; ++i)
    {
        double f_x_est = f(x_est);
        double f_df = f_x_est / df(x_est);
        if ((right - (x_est - f_df)) * ((x_est - f_df) - left) < 0.0 || std::fabs(f_df) < (right - left) / 4)
        {
            if (f_x_est > 0)
            {
                right = x_est;
            }
            else
            {
                left = x_est;
            }
            f_df = x_est - (right + left) / 2;
        }
        x_est -= f_df;
        if (std::fabs(f_df) < eps)
        {
            break;
        }
    }
    return x_est;
}

std::vector<double> cost_optimal(Vector2d &x0, Vector2d &v0, Vector2d &x1, Vector2d &v1)
{
    Vector2d x01 = x1 - x0;
    Vector2d v01 = v1 - v0;

    double p = -4 * (v0.dot(v0) + v1.dot(v1) + v0.dot(v1));
    double q = 24 * x01.dot(v0 + v1);
    double r = -36 * x01.dot(x01);

    double t_min = 0, t_max = 10;
    auto cost = [=](double t) -> double {
        return t + v01.dot(4 * v01 / t - 6 * (-t * v0 + x01) / (std::pow(t, 2))) + (-t * v0 + x01).dot(-6 * v01 / std::pow(t, 2) + 12 * (x01 - v0 * t) / std::pow(t, 3));
    };
    auto d_cost = [=](double t) -> double { return std::pow(t, 4) + p * t * t + q * t + r; };
    auto dd_cost = [=](double t) -> double { return 4.0 * std::pow(t, 3) + 2 * p * t + q; };

    double t_star = bisection_newton(d_cost, dd_cost, t_min, t_max);
    double cost_t = cost(t_star);
    return std::vector<double>{cost_t, t_star};
}

std::vector<double> cost_optimal(const std::vector<double> &s0, const std::vector<double> &s1)
{
    Vector2d a, b, c, d;
    a << s0[0], s0[1];
    b << s0[2], s0[3];
    c << s1[0], s1[1];
    d << s1[2], s1[3];
    return cost_optimal(a, b, c, d);
}

std::vector<double> forward_reachable_box(Vector2d &x0, Vector2d &v0, double r)
{
    Vector2d tau_x_plus = (2. / 3) * (-v0.array().pow(2) + r + v0.array() * (v0.array().pow(2) + r).sqrt()).matrix();
    Vector2d tau_x_minus = (2. / 3) * (-v0.array().pow(2) + r - v0.array() * (v0.array().pow(2) + r).sqrt()).matrix();

    Vector2d xmax = (v0.array() * tau_x_plus.array() + x0.array() + ((1. / 3) * tau_x_plus.array().pow(2) * (-tau_x_plus.array() + r)).sqrt()).matrix();
    Vector2d xmin = (v0.array() * tau_x_minus.array() + x0.array() - ((1. / 3) * tau_x_minus.array().pow(2) * (-tau_x_minus.array() + r)).sqrt()).matrix();

    double tau_v_plus = 0.5 * r;
    Vector2d vmax = (v0.array() + std::sqrt(tau_v_plus * (-tau_v_plus + r))).matrix();
    Vector2d vmin = (v0.array() - std::sqrt(tau_v_plus * (-tau_v_plus + r))).matrix();
    std::vector<double> ret{xmin(0), xmax(0), xmin(1), xmax(1), vmin(0), vmax(0), vmin(1), vmax(1)};
    return ret;
}

std::vector<double> backward_reachable_box(Vector2d &x0, Vector2d &v0, double r)
{
    Vector2d tau_x_plus = (2. / 3) * (-v0.array().pow(2) - r + v0.array() * (v0.array().pow(2) + r).sqrt()).matrix();
    Vector2d tau_x_minus = (2. / 3) * (-v0.array().pow(2) - r - v0.array() * (v0.array().pow(2) + r).sqrt()).matrix();

    Vector2d xmax = (v0.array() * tau_x_plus.array() + x0.array() + ((1. / 3) * tau_x_plus.array().pow(2) * (-tau_x_plus.array() + r)).sqrt()).matrix();
    Vector2d xmin = (v0.array() * tau_x_minus.array() + x0.array() - ((1. / 3) * tau_x_minus.array().pow(2) * (-tau_x_minus.array() + r)).sqrt()).matrix();

    double tau_v_plus = 0.5 * r;
    Vector2d vmax = (v0.array() + std::sqrt(tau_v_plus * (tau_v_plus + r))).matrix();
    Vector2d vmin = (v0.array() - std::sqrt(tau_v_plus * (tau_v_plus + r))).matrix();
    std::vector<double> ret{xmin(0), xmax(0), xmin(1), xmax(1), vmin(0), vmax(0), vmin(1), vmax(1)};
    return ret;
}

bool isinside(const std::vector<double> &recBox, const std::vector<double> &s)
{
    return (recBox[0] < s[0] and s[0] < recBox[1]) and (recBox[2] < s[1] and s[1] < recBox[3]) and (recBox[4] < s[2] and s[2] < recBox[5]) and (recBox[6] < s[3] and s[3] < recBox[7]);
}

reachFilter filter_reachable(const std::vector<std::vector<double>> &Sset, const std::vector<int> &idxset, std::vector<double> &s_c, double r, bool ForR)
{
    Vector2d x_c, v_c;
    x_c << s_c[0], s_c[1];
    v_c << s_c[2], s_c[3];

    std::vector<double> recBox = ForR ? forward_reachable_box(x_c, v_c, r) : backward_reachable_box(x_c, v_c, r);
    std::vector<int> idx_filter;
    std::vector<double> dist_filter, time_filter;
    for (const auto &idx : idxset)
    {
        if (isinside(recBox, Sset[idx]))
        {
            std::vector<double> cost_tau = ForR ? cost_optimal(s_c, Sset[idx]) : cost_optimal(Sset[idx], s_c);
            if (cost_tau[0] < r)
            {
                idx_filter.push_back(idx);
                dist_filter.push_back(cost_tau[0]);
                time_filter.push_back(cost_tau[1]);
            }
        }
    }
    return std::make_tuple(idx_filter, dist_filter, time_filter);
}

std::vector<std::vector<double>> gen_trajectory(const std::vector<double> &s0, const std::vector<double> &s1, double tau, int N_split)
{
    std::vector<std::vector<double>> waypoints;
    if (tau < 0.001)
    {
        return waypoints;
    }
    Vector4d s0M, s1M;
    s0M << s0[0], s0[1], s0[2], s0[3];
    s1M << s1[0], s1[1], s1[2], s1[3];
    Vector2d x0, v0, x1, v1, x01, v01;
    x0 << s0[0], s0[1];
    v0 << s0[2], s0[3];
    x1 << s1[0], s1[1];
    v1 << s1[2], s1[3];
    x01 = x1 - x0;
    v01 = v1 - v0;

    Vector4d d;
    Vector2d du, dd; // d up, d down
    du = -6 * v01 / (tau * tau) + 12 * (-tau * v0 + x01) / (std::pow(tau, 3));
    dd = 4 * v01 / tau - 6 * (-tau * v0 + x01) / (tau * tau);
    d << du(0), du(1), dd(0), dd(1);

    for (int i = 1; i < N_split + 2; ++i)
    {
        double t = (i - 1) * tau / N_split;
        double s = t - tau;
        Matrix2d eye = Matrix2d::Identity();

        Matrix4d M_left, M_right;
        M_left << eye, eye * s,
            eye * 0, eye;
        M_right << eye * (-std::pow(s, 3) * (1. / 6)), eye * (std::pow(s, 2)) * 0.5,
            eye * (-0.5 * std::pow(s, 2)), eye * s;
        Vector4d waypointM;
        waypointM = M_left * s1M + M_right * d;
        waypoints.push_back(std::vector<double>{waypointM[0], waypointM[1], waypointM[2], waypointM[3]});
    }
    return waypoints;
}

namespace plt = matplotlibcpp;

void show_trajectory(const std::vector<double> &s0, const std::vector<double> &s1, double tau, int N_split, std::string color, std::string linewidth)
{
    if (tau < 0.001)
    {
        return;
    }
    std::vector<std::vector<double>> waypoints = gen_trajectory(s0, s1, tau, N_split);
    // 绘出结果
    std::vector<double> x, y;
    for (const auto &state : waypoints)
    {
        x.push_back(state[0]);
        y.push_back(state[1]);
    }
    std::map<std::string, std::string> style;
    style.insert(std::make_pair("c", color));
    style.insert(std::make_pair("linewidth", linewidth));
    plt::plot(x, y, style);
}

File: fmt/pythonfmt/doubleintegrator.py
import numpy as np
import random
import matplotlib as mpl
import matplotlib.pyplot as plt
import time


def bisection_newton(f, df, left, right, eps=0.05, itr_max=20):
    x_est = (left + right) * 0.5
    for itr in range(1, itr_max):
        f_df = f(x_est) / df(x_est)
        if (right - (x_est - f_df)) * ((x_est - f_df) - left) < 0.0 or abs(f_df) < (right - left) / 4.0:
            if f(x_est) > 0:
                right = x_est
            else:
                left = x_est
            f_df = x_est - (right + left) * 0.5
        x_est -= f_df
        if abs(f_df) < eps:
            break
    return x_est


def cost_optimal_in(x0, v0, x1, v1):
    # print("x0", x0, "v0", v0, "x1", x1, "v1", v1)
    x01 = x1 - x0
    v01 = v1 - v0
    # print("x01", x01, "v01", v01)
    p = -4 * (np.sum(v0 * v0) + np.sum(v1 * v1) + np.sum(v0 * v1))
    q = 24 * np.sum((v0 + v1) * x01)
    r = -36 * np.sum(x01 * x01)

    # print("p", p, "q", q, "r", r)
    def cost(t):
        return t + np.sum(v01 * (4.0 * v01 / t - 6 * (-v0 * t + x01) / t ** 2)) + np.sum(
            (-6 * v01 / np.power(t, 2) + 12 * (x01 - v0 * t) / np.power(t, 3)) * (-v0 * t + x01))

    def d_cost(t):
        return np.power(t, 4) + p * t * t + q * t + r  # df(t)/dt

    def dd_cost(t):
        return 4.0 * t * t * t + 2 * p * t + q  # ddf(t)/dt^2

    t_min = 0.0
    t_max = 10.0
    t_star = bisection_newton(d_cost, dd_cost, t_min, t_max)
    # print("t_star", t_star)
    return cost(t_star), t_star


def cost_optimal(s0, s1):
    a = s0[0:2]
    b = s0[2:4]
    c = s1[0:2]
    d = s1[2:4]
    return cost_optimal_in(a, b, c, d)


def forward_reachable_box(x0, v0, r):
    tau_x_plus = (2. / 3) * (-np.power(v0, 2) + r + v0 * np.sqrt(np.power(v0, 2) + r))
    tau_x_minus = (2. / 3) * (-np.power(v0, 2) + r - v0 * np.sqrt(np.power(v0, 2) + r))

    xmax = v0 * tau_x_plus + x0 + np.sqrt((1. / 3) * np.power(tau_x_plus, 2) * (-tau_x_plus + r))
    xmin = v0 * tau_x_minus + x0 - np.sqrt((1. / 3) * np.power(tau_x_minus, 2) * (-tau_x_minus + r))

    tau_v_plus = 0.5 * r
    vmax = v0 + np.sqrt(np.dot(tau_v_plus, (-tau_v_plus + r)))
    vmin = v0 - np.sqrt(np.dot(tau_v_plus, (-tau_v_plus + r)))
    return xmin, xmax, vmin, vmax


def backward_reachable_box(x0, v0, r):
    tau_x_plus = (2. / 3) * (np.power(v0, 2) - r + v0 * np.sqrt(np.power(v0, 2) + r))
    tau_x_minus = (2. / 3) * (np.power(v0, 2) - r - v0 * np.sqrt(np.power(v0, 2) + r))
    xmax = v0 * tau_x_plus + x0 + np.sqrt((1. / 3) * np.power(tau_x_plus, 2) * (tau_x_plus + r))
    xmin = v0 * tau_x_minus + x0 - np.sqrt((1. / 3) * np.power(tau_x_minus, 2) * (tau_x_minus + r))

    tau_v_plus = 0.5 * r
    vmax = v0 + np.sqrt(np.dot(tau_v_plus, (tau_v_plus + r)))
    vmin = v0 - np.sqrt(np.dot(tau_v_plus, (tau_v_plus + r)))
    return xmin, xmax, vmin, vmax


def filter_reachable(Sset, idxset, s_c, r, ForR):
    # s_set_filtered = []
    x_c = s_c[0:2]
    v_c = s_c[2:4]
    # print('x_c', x_c, 'v_c', v_c)
    xmin, xmax, vmin, vmax = forward_reachable_box(x_c, v_c, r) if ForR == "F" else backward_reachable_box(x_c, v_c, r)

    # print('xmin', xmin, 'xmax', xmax, 'vmin', vmin, 'vmax', vmax)

    def isinside(s):
        if xmin[0] < s[0] < xmax[0] and xmin[1] < s[1] < xmax[1] and vmin[0] < s[2] < vmax[0] and vmin[1] < s[3] < vmax[
            1]:
            return True
        else:
            return False

    idx_filter = []
    dist_filter = []
    time_filter = []
    for idx in idxset:
        s = Sset[idx]
        if isinside(s):
            cost, tau = cost_optimal(s, s_c) if ForR == "B" else cost_optimal(s_c, s)
            # print("cost", cost, "tau", tau)
            if cost < r:
                idx_filter.append(idx)
                dist_filter.append(cost)
                time_filter.append(tau)
    return idx_filter, dist_filter, time_filter


def gen_trajectory(s0, s1, tau, N_split=10):
    if tau == 0:
        return np.array([])
    x0 = s0[0:2]
    v0 = s0[2:4]
    x1 = s1[0:2]
    v1 = s1[2:4]
    # print("x0", x0, "v0", v0, "x1", x1, "v1", v1)
    x01 = x1 - x0
    v01 = v1 - v0
    # print("x01", x01, "v01", v01)
    d = np.concatenate(
        (-6 * v01 / tau ** 2 + 12 * (-tau * v0 + x01) / tau ** 3, 4 * v01 / tau - 6 * (-tau * v0 + x01) / tau ** 2))

    # print("d", d)
    def f(t):
        s = t - tau
        eye = np.identity(2)
        M_left = np.concatenate((np.concatenate((eye, eye * s), axis=1), np.concatenate((eye * 0, eye), axis=1)),
                                axis=0)
        # print("M_left",M_left)
        M_right = np.concatenate((np.concatenate((eye * (-s ** 3) * (1.0 / 6.0), eye * s ** 2 * 0.5), axis=1),
                                  np.concatenate((eye * (-s ** 2 * 0.5), eye * s), axis=1)), axis=0)
        # print("M_right", M_right)
        # print(d.shape)
        # print("res", [np.dot(M_left, s1)])
        return np.dot(M_left, s1.T) + np.dot(M_right, d.T)

    waypoints = []
    for n in range(1, N_split + 2):
        t = (n - 1) * tau / N_split
        waypoints.append(f(t))

    return np.array(waypoints)


def show_trajectory(s0, s1, tau, N_split=20, ax=None, c='gray', linewidth_=0.2):
    if tau == 0:
        return

    waypoints = gen_trajectory(s0, s1, tau, N_split)
    M = np.zeros((N_split + 1, 4))
    for i in range(0, N_split + 1):
        # print(waypoints[i])
        M[i, :] = waypoints[i]
    ax.plot(M[:, 0], M[:, 1], c=c, linewidth=linewidth_)
    # plt.show()


def test():
    N = 10 ** 5
    s_set = np.empty([N, 4])
    for i in range(0, N):
        s_set[i, :] = [random.random() - 0.5, random.random() - 0.5, 3 * random.random() - 0.5,
                       3 * random.random() - 0.5]

    s_c = np.array([0, 0, 0.2, 0.2])
    idxset = [x for x in range(0, N)]

    idx_for, trash, _ = filter_reachable(s_set, idxset, s_c, 1.2, "F")
    idx_back, trash, _ = filter_reachable(s_set, idxset, s_c, 1.2, "B")
    s_for = []
    for idx in idx_for:
        s_for.append(s_set[idx])
    s_back = []
    for idx in idx_back:
        s_back.append(s_set[idx])

    setA = np.zeros((2, len(idx_for)))
    setB = np.zeros((2, len(idx_back)))
    for i in range(0, len(idx_for)):
        setA[:, i] = s_for[i][0:2]

    for i in range(0, len(idx_back)):
        setB[:, i] = s_back[i][0:2]

    fig, ax = plt.subplots()  # Create a figure and an axes.
    ax.scatter(setA[0], setA[1])
    ax.scatter(setB[0], setB[1], marker='x')
    plt.show()
    return setA


def test_connection():
    """
    s0 = Vec4f(0.17, 0.08, 0.19, -0.06)
    s1 = Vec4f(0.22, 0.03, 0.139, -0.2719)
    """
    s0 = np.array([0.0, 0.0, 0.3, 0.0])
    s1 = np.array([0.1, 0.1, -0.1, -0.01])
    fig, ax = plt.subplots()  # Create a figure and an axes.
    show_trajectory(s0, s1, 0.5, ax=ax)
    plt.show()


if __name__ == "__main__":
    start = time.time()
    test()
    test_connection()
    end = time.time()
    print(end - start)
